### DISCUSSION TITLE
Move JavaScript runtime toward ES6 and add jsdoc annotations

### DISCUSSION THREAD

**Discussion by jcking** [antlr/antlr4#3411] [General]
The JavaScript runtime would benefit from moving toward ES6 and adding jsdoc annotations. To do this Map and Set in `Utils.js` would need to be made compatible with ES6 Map and Set. Then we would detect if Map and Set is available on the global object and use those instead. By adding jsdoc annotations, the TypeScript compiler can generate type declarations allowing better interop with TypeScript.

**Comment by ericvergnaud** [comment]
The JS runtime is already ES6 compatible, so not sure what you mean by "reward" ? 
ES6 Map and Set have a behavior different from the one required by ANTLR, so that's not going to change.
And Typescript is an existing different target, just not supported by the core team.

**Reply by jcking** [reply]
It does not use ES6 modules (import and export) and instead uses `require`. By adding jsdoc annotations, there would be no reason to require a separate runtime target. Only a separate language generation target would be needed, as the JavaScript runtime could be used for both.
